subprojects/execution/src/main/java/org/gradle/internal/execution/steps/AbstractValidateStep.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.steps;
+
+import com.google.common.collect.ImmutableCollection;
+import com.google.common.collect.ImmutableSet;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.execution.WorkValidationContext;
+import org.gradle.internal.execution.WorkValidationException;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.reflect.validation.TypeValidationProblem;
+import org.gradle.internal.reflect.validation.TypeValidationProblemRenderer;
+import org.gradle.model.internal.type.ModelType;
+
+import java.util.Collection;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Abstract class for steps that validate a {@link UnitOfWork} or report validation errors present in the {@link WorkValidationContext}.
+ *
+ * @param <C> previous execution context
+ * @param <R> result of execution delegate
+ */
+/* package */ abstract class AbstractValidateStep<C extends Context, R extends Result> extends BuildOperationStep<C, R> {
+    private static final int MAX_ERR_COUNT = Integer.getInteger("org.gradle.internal.max.validation.errors", 5);
+
+    public AbstractValidateStep(BuildOperationExecutor buildOperationExecutor) {
+        super(buildOperationExecutor);
+    }
+
+    protected void throwValidationException(UnitOfWork work, WorkValidationContext validationContext, Collection<TypeValidationProblem> validationErrors) {
+        ImmutableSet<String> uniqueErrors = validationErrors.stream()
+                .map(TypeValidationProblemRenderer::renderMinimalInformationAbout)
+                .collect(ImmutableSet.toImmutableSet());
+        throw WorkValidationException.forProblems(uniqueErrors)
+                .limitTo(MAX_ERR_COUNT)
+                .withSummary(new Summarizer(work, validationContext))
+                .get();
+    }
+
+    private static class Summarizer implements Function<WorkValidationException.Builder.SummaryHelper, String> {
```

Comment:
I don't see much value in extracting this into its own class. Now we have to have type parameters pointing to deeply nested classes as well as fields retaining state for local variables. All this is done by the compiler when you use the original notation of `.withSummary(helper -> ...)`. I think we should stick with that.

Reply:
The type parameter using the deeply nested classes is admittedly quite ugly.  I don't see why final fields vs. locals is a concern for a short-lived class that will only exist in the context of "we're failing, an exception is about to be thrown" though?

I think there is a significant benefit in separating the two concerns of 1) performing validation and 2) building an error message summary.  This class hides all the details of how to build the summary from the `BuildOperationStep` extension using it.  Being a step and doing validation seems like enough responsibility for that class.  This class makes it easier to read both this code and the step class using it by isolating a smaller area of concern for a not-super-complex but certainly non-trivial formatting process.  The summarization logic arguably might also benefit from a unit test, which this could enable.

How strongly do you feel about this?  I really think even with the ugly generics, this refactoring is an improvement I'd like to keep as I rework this PR.

Reply:
I ended up moving this and adding an import to hide the generic ugliness.  I'm also unsure why `ValidatePlugins.validateTaskClasses()` does not limit the max errors.  Is this intentional?  Maybe we can make that limit a default set during construction of the builder and further simplify creating these exceptions?

Reply:
+1 on making the limit the default. If we have some use-cases where we don't want to limit the number, then the builder could have something like `reportAll()` or similar.